dds_web/api/schemas/project_schemas.py

```
@@ -84,3 +86,120 @@ def return_items(self, data, **kwargs):
+
+        # Check if project has contents
+        project_row = verify_project_exists(spec_proj=data.get("project"))
+        if not project_row.files:
+            raise ddserr.EmptyProjectException
+
+        # Check if specific files have been requested or if requested all contents
+        files, folder_contents, not_found = (None, None, None)
+        if requested_items:
+            files, folder_contents, not_found = self.find_contents(
+                project=project_row, contents=requested_items
+            )
+        elif get_all:
+            files = project_row.files
+        else:
+            raise ddserr.DDSArgumentError(message="No items were requested.")
+
+        # Items to return
+        found_files = {}
+        found_folder_contents = {}
+        not_found = {}
+
+        # Use file schema to get file info automatically
+        fileschema = sqlalchemyautoschemas.FileSchema(
+            many=False,
+            only=(
+                "name_in_bucket",
+                "subpath",
+                "size_original",
+                "size_stored",
+                "salt",
+                "public_key",
+                "checksum",
+                "compressed",
+            ),
+        )
+
+        # Connect to s3
+        with api_s3_connector.ApiS3Connector(project=project_row) as s3:
+            # TODO: Optimisation: Add check for if only searching for one file (head_bucket)
+            # Get bucket items
+            pages = s3.bucket_items()
+
+            # Get the info and signed urls for all files found in the bucket
+            for page in pages:
+                found_files.update(
+                    {
+                        x.name: {
+                            **fileschema.dump(x),
+                            "url": s3.generate_get_url(key=x.name_in_bucket) if url else None,
+                        }
+                        for x in files
+                        if x.name_in_bucket in page
+                    }
```

If I understand this correctly, I think we can remove the check if a file is in the bucket or not. If a file that is in the database is not found in the bucket, that's a fairly serious error and we should raise an exception instead of silently filtering them out of the download list. I think you said the http request in the presigned URL would return 403 for non-existing files so I guess it would be easiest to raise that exception in the CLI for the user.

---

Ok! I was thinking of doing something like that originally, I have no objections, will change.

---

Done